### PR TITLE
Keep a record of historically submitted jobs

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The old `Submission` class has been refactored into a new `Job` model. These `jobs` are now persisted as a `metadata.yaml` file within the root filesystem. This allows the user's to get a list of previous jobs. The new routes are:

* GET /v0/history/jobs - Index all old jobs
* GET /v0/history/jobs/:id - Get the information about a single job

The old `scheduler_command` has been removed and replaced with scheduler specific scripts. The new `submit.sh` script is called when a job is submitted and `monitor.sh` is used to poll for data about the `job`. The file both depends on the specified `scheduler` environment variable:

```
libexec/<scheduler>/submit.sh
libexec/<scheduler/monitor.sh
```

Both the `submit.sh` and `monitor.sh` feed data back into the server process. This is done by emitting a JSON payload as the last line to STDOUT. This requires some boilerplate code to format the response using `jq`.

To make it easier to modify on a per installation basis, the `slurm/submit.sh` sub-executes `slurm/sbatch.sh`. This allows the `sbatch.sh` to be modified whist the communication protocol is handled by `submit.sh`. The `monitor.sh` does not have an equivalent feature as it simply wraps the output of `scontrol show job`.

The `monitor.sh` is executed immediately after a job is submitted, when a job is retrieved explicitly by `id`, and periodically. This is a compromise between giving accurate information on demand and ensure the data is retrieved before `scontrol` clears its cache.

The following fields have been implemented for the jobs:
* `created_at`
* `scheduler_id`
* `state`
* `reason`
* `stdout_path`
* `stderr_path`
* `start_time` (actual/predicated depending on the `state`)
* `end_time` (actual/predicated depending on the `state`)

Based on #17

### CODE REVIEW

**Comment by WilliamMcCumstie** [comment]
Rebased on master

**Comment by WilliamMcCumstie** [comment]
Rebased to get the latest version of the config

**Comment by WilliamMcCumstie** [comment]
Update `GET /render/:id` to `POST /render/:id` in documentation before merging

**Comment by benarmston** [comment]
Rebased on `master`

**Comment by benarmston** [comment]
I've made my first pass at trying to write the client code against this PR.  I've ran into the following issues:

1. I want to display details of the script and template that the job originates from.  I want to be able to `?include=script,script.template` but those includes are not supported.
2. The metadata file contains useful information for why a job could not be submitted, namely `submit_stderr`, but the job API doesn't expose it.  Can you add both `submit_stderr` and `submit_stdout`?

If solving (1) above requires the separation between the history and template apps to be removed.  I'm OK with that.  I think that this highlights that they are probably the same concern, not separate concerns as I'd originally considered them to be.

**Comment by benarmston** [comment]
Rebased on `master`.

File: api/lib/flight_job_script_api/system_command.rb [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      @cmd = cmd
+      @user = user
+      @stdin = stdin
+      @mutex = self.class.mutexes[user]
+    end
+
+    def wait
+      return if @wait
+      run
+      @wait ||= true
```
Comment (benarmston) [inline api/lib/flight_job_script_api/system_command.rb:91]: This looks ood.  What could `@wait` be set to here that doesn't result in it being set to `true`?

File: api/libexec/slurm/monitor.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#-------------------------------------------------------------------------------
+# WARNING - README
+#
+# This is an internally managed file, any changes maybe lost on the next update!
+# Please clone the entire 'slurm' directory in order to modify this file.
+#-------------------------------------------------------------------------------
+
+# Ensure jq is on the path
+set -e
+which "jq"
```
Comment (benarmston) [inline api/libexec/slurm/monitor.sh:38]: Lets silence this in the good case.

File: api/libexec/slurm/monitor.sh [hunk 3]
```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
```
Comment (benarmston) [inline api/libexec/slurm/monitor.sh:1]: Is this file still needed?  Is it not part of the CLI now?

Comment (benarmston) [reply api/libexec/slurm/monitor.sh:1]: Same question for all the files in the `libexec` directory.

**Review by benarmston — CHANGES_REQUESTED** [review]
A few comments for you to look at @WilliamMcCumstie

File: api/app.rb [hunk 4]
```diff
@@ -236,30 +217,39 @@ class RenderApp < Sinatra::Base
...
-        halt
-      end
+    # Transform any multiselect answers from CSV to Arrays
+    answers = params.to_h.dup
+    template.generation_questions.each do |q|
+      next unless q.metadata['format']['type'] == 'multiselect'
+      next unless answers[q.id].is_a? String
+      FlightJobScriptAPI.logger.warn "Implicitly converting '#{q.id}' from CSV to an array"
+      answers[q.id] = CSV.parse(answers[q.id]).first
+    end
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app.rb:235]: Yep, the CSV translation exists because of how the client operates. Initially when I implemented the CLI, I didn't realise it was in the CSV format as opposed to atomic values. So the CLI when down the path of arrays, which in many ways is easier.

However this has left us with a disconnect between arrays in the CLI and CSV in the web client. There is another translation in the serialisation's of the the questions' defaults as document here:
https://github.com/openflighthpc/flight-job-script-service/blob/dev/active-jobs/api/docs/routes.md#get---templatesidquestions

If you would prefer the client worked with `arrays`, then all the CSV handling could be removed.

File: api/app/serializers/question_serializer.rb [hunk 5]
```diff
@@ -27,5 +27,14 @@
...
+  ['text', 'description', 'default', 'format', 'ask_when'].each do |field|
+    attribute(field) { object.metadata[field] }
+  end
+
+  # NOTE: Technically only multiselect questions have Array defaults, however
+  # it is "assumed" this is enforced by the underlying CLI
+  attribute(:default) do
+    default = object.metadata['default']
+    default.is_a?(Array) ? CSV.generate_line(default).chomp : default
+  end
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app/serializers/question_serializer.rb:39]: Consistency with here:

https://github.com/openflighthpc/flight-job-script-service/pull/18#discussion_r598666196

File: api/lib/flight_job_script_api/system_command.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply api/lib/flight_job_script_api/system_command.rb:91]: We don't have any form of promises in the `SystemCommand` class, so the `wait` method takes its place. Its purposes is to ensure the system command has been ran before returning, but also ensures it runs only once.

The actual value of `@wait` doesn't really matter, apart from the fact it is truthy.

Comment (benarmston) [reply api/lib/flight_job_script_api/system_command.rb:91]: It's the use of `||=` rather than `=` that confuses me.  It suggests that `@wait` could be set to something truthy inside of `run` and that we want to preserve that value.

File: api/app.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (benarmston) [reply api/app.rb:235]: I think using JSON arrays would be simpler.  Can you make the changes to the API and I'll update the client.

File: api/libexec/slurm/monitor.sh (hunk 3 above)
Comment (WilliamMcCumstie) [reply api/libexec/slurm/monitor.sh:1]: e0821664dbe19fed298c0c37846652b20d0e34fa

Comment (WilliamMcCumstie) [reply api/libexec/slurm/monitor.sh]: e0821664dbe19fed298c0c37846652b20d0e34fa

File: api/lib/flight_job_script_api/configuration.rb [hunk 6]
```diff
@@ -83,5 +84,36 @@ class ConfigError < StandardError; end
...
+
+    def submit_script_path
+      @submit_script_path ||= File.join(script_dir, scheduler, 'submit.sh')
+      @submit_script_path.tap do |path|
+        unless File.exists?(path)
+          raise ConfigError, <<~ERROR
+            The submit.sh script does not exist: #{path}
+          ERROR
+        end
+        unless [1, 5, 7].include?(File.stat(path).mode % 10)
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/lib/flight_job_script_api/configuration.rb]: facf2521cce49d4396f340618eab15ff5608bba7

File: api/lib/flight_job_script_api/configuration.rb [hunk 7]
```diff
@@ -83,5 +84,36 @@ class ConfigError < StandardError; end
...
+    end
+
+    def monitor_script_path
+      @monitor_script_path ||= File.join(script_dir, scheduler, 'monitor.sh').tap do |path|
+        unless File.exists?(path)
+          raise ConfigError, <<~ERROR
+            The monitor.sh script does not exist: #{path}
+          ERROR
+        end
+        unless [1, 5, 7].include?(File.stat(path).mode % 10)
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/lib/flight_job_script_api/configuration.rb]: facf2521cce49d4396f340618eab15ff5608bba7

File: api/lib/flight_job_script_api/system_command.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply api/lib/flight_job_script_api/system_command.rb]: b51cc0e265f8829ae2aafa56ff0f1975eca4953d

File: api/app.rb [hunk 8]
```diff
@@ -151,50 +127,55 @@ def find(id)
...
+
+    show
+
     create do |attr|
-      sub = Submission.new
-      [sub.id, sub]
+      @action = :create
+      # Due to the how the internal Sinja routing works, the job needs an "ID"
+      # However the actual ID won't be assigned until latter, so a temporary ID
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app.rb]: 56bfe5664583b390ddf4ebc25021ef94be92a727

File: api/app/models/job.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+      script_id = metadata['script_id']
+      FlightJobScriptAPI.logger.info "Lazy loading related script: #{script_id} (script: #{id})"
+      @script = Script.find(script_id, user: user)
+    end
+    @script
+  end
+
+  def submit
+    unless script_id
+      raise MissingScript, "Can not create a job without a script"
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app/models/job.rb]: 56bfe5664583b390ddf4ebc25021ef94be92a727

File: api/app.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app.rb]: 1930e1df42c47c8ce1354601b22f497800aab7e0

File: api/app/serializers/question_serializer.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app/serializers/question_serializer.rb]: 1930e1df42c47c8ce1354601b22f497800aab7e0

**Comment by benarmston** [comment]
Thanks for your work on this @WilliamMcCumstie
